Bio/bgzf.py

```diff
@@ -852,33 +871,58 @@ def _write_block(self, block):
         data = _bgzf_header + bsize + compressed + crc + uncompressed_length
         self._handle.write(data)
 
+    def _write_bytes(self, data):
+        self._buffer += data
+        while len(self._buffer) >= self._BLOCK_SIZE:
+            self._write_block(self._buffer[: self._BLOCK_SIZE])
+            self._buffer = self._buffer[self._BLOCK_SIZE :]
+
+    def _write_str(self, data_str):
+        assert isinstance(data_str, str)
+        data_bytes = data_str.encode(encoding=self._encoding)
+        if len(data_bytes) == len(data_str):
+            # fast case -- 8-bit only
+            self._write_bytes(data_bytes)
+            return
+
+        # When reading we can't cope with multi-byte UTF-8 characters
+        # being split between BGZF blocks. Therefore, if the 64KB block
+        # end "in the middle" of a character, the block is resized down
+        # to the end of the previous character. This can produce block
+        # up to 3 bytes smaller than the full 64KB.
+        assert self._encoding == "utf-8"
+        assert len(self._buffer) < self._BLOCK_SIZE
+        while len(data_bytes):
+            safe_size = self._BLOCK_SIZE - len(self._buffer)
+            if safe_size >= len(data_bytes):
+                safe_size = len(data_bytes)
+            else:
+                while safe_size > 0 and data_bytes[safe_size] >> 6 == 2:
+                    # the byte after the block boundary is not the first octet
+                    # of UTF-8 character (RFC 3629)
+                    safe_size -= 1
```

Comment:
This is the magic (the key bit of logic to prevent multi-byte characters being split).

https://www.rfc-editor.org/rfc/rfc3629 says "Character boundaries are easily found from anywhere in an octet stream." but I don't immediately see how.

Presumably this is only for utf-8, but will it cope with all the weird and wonderful things in unicode? e.g. emoji and zero-width-joins?